llama_index/node_parser/postprocessors/metadata_extractors.py

```diff
@@ -0,0 +1,393 @@
+                  appropriate __call__ or a simply a function/lambda function.
+        """
+        self._extractors.append(extractor)
+        return self
+
+    def build(self) -> "MetadataAugmentationPostProcessor":
+        return MetadataAugmentationPostProcessor(
+            self._extractors, self._node_text_template, self._disable_template_rewrite
+        )
+
+
+class MetadataAugmentationPostProcessor(NodeParserPostProcessor):
+    def __init__(
+        self,
+        extractors: Sequence[Callable[[Sequence[BaseNode]], None]],
+        node_text_template: Optional[str] = None,
+        disable_template_rewrite: bool = False,
+    ) -> None:
+        self._extractors = extractors
+        self._node_text_template = (
+            node_text_template
+            or """\
+            [Excerpt from document]\n{metadata_str}\n\
+            Excerpt:\n-----\n{content}\n-----\n"""
+        )
+        self._disable_template_rewrite = disable_template_rewrite
+
+    def post_process_nodes(self, nodes: Sequence[BaseNode]) -> None:
+        """Extract metadata from a document.
+
+        Args:
+            nodes (Sequence[BaseNode]): nodes to extract metadata from
+
+        """
+        if not self._disable_template_rewrite:
+            for node in nodes:
+                if isinstance(node, TextNode):
+                    cast(TextNode, node).text_template = self._node_text_template
+        for extractor in self._extractors:
+            extractor(nodes)
+
+
+class TitleExtractor(MetadataExtractorBase):
+    """Title extractor. Useful for long documents. Extracts `document_title`
+    metadata field.
+    Args:
+        nodes (int): number of nodes from front to use for title extraction
+        node_template (str): template for node-level title clues extraction
+        combine_template (str): template for combining node-level clues into
+            a document-level title
+    """
+
+    def __init__(
+        self,
+        llm_predictor: BaseLLMPredictor,
+        nodes: int = 5,
+        node_template: Optional[str] = None,
+        combine_template: Optional[str] = None,
+    ) -> None:
+        """Init params."""
+        if nodes < 1:
+            raise ValueError("num_nodes must be >= 1")
+        self._nodes = nodes
+        self._node_template = node_template
+        self._combine_template = combine_template
+        self._llm_predictor = llm_predictor
+
+    def __call__(self, nodes: Sequence[BaseNode]) -> None:
+        nodes_to_extract_title: List[BaseNode] = []
+        for node in nodes:
+            if len(nodes_to_extract_title) >= self._nodes:
+                break
+            if self.is_text_node_only and not isinstance(node, TextNode):
+                continue
+            nodes_to_extract_title.append(node)
+
+        if len(nodes_to_extract_title) == 0:
+            # Could not extract title
+            return
+
+        title_candidates = [
+            self._llm_predictor.predict(
```

Comment:
i'd consider using the ResponseSynthesizer module instead, it has in-built capabilities to abstract over long contexts

Reply:
If this is switched, seems like a good use case for the accumulate synthesizer

Reply:
I suggest to utilize more of the existing prompt infra in a follow up PR so that we can get this functional feature released / publicized sooner.